### ISSUE TITLE
AHK-like functionality on MacOS via Hammerspoon

### ISSUE THREAD

**Issue by jessecrouch** [fwextensions/QuicKey#8]
Hi -

I think there might be a way to do the extended functionality that QuicKey has on Windows (via AHK) while on Mac.

I'm using [Hammerspoon](https://www.hammerspoon.org/) to implement this script that was originally used for the CLUT extension: https://stackoverflow.com/questions/39233870/mru-browser-tab-switching-with-hammerspoon-luamac-issue

```lua
-- MRU for QuicKey
local ctrlTab = hs.hotkey.new({"ctrl"}, "tab", nil, function()
  hs.eventtap.keyStroke({"ctrl"}, "q")
end)
chromeWatcher = hs.application.watcher.new(function(name, eventType, app)
  if eventType ~= hs.application.watcher.activated then return end
  if name == "Brave Browser" then
    ctrlTab:enable()
  else
    ctrlTab:disable()
  end
end)
```

I'm able to trigger the ctrl+q action using ctrl+tab, but I'm not sure how to clone the rest of the AHK script. Any help appreciated.

**Comment by fwextensions** [comment]
Interesting.  I've never tried Hammerspoon, and don't have a Mac for easy testing.  I actually adapted the AHK code from someone else's script, so I'm not an AHK expert.  But my understanding of what it's doing is:

	on ctrl-tab down:
		sawCtrlTab = true
		if the QK popup isn't open:
			send ctrl-Q
			loop every 100ms:
				if !sawCtrlTab:
					exit // got a ctrl up while we were sleeping
				else if popup is active:
					sleep 300ms // give the menu JS time to init so the first down arrow isn't lost
					send down arrow // select the first tab in the list so if you release ctrl now, it'll switch to the previous tab
		else:
			send down arrow // to select the next item

	on ctrl-shift-tab down:
		if the QK popup isn't open:
			send ctrl-Q
		else: 
			send up arrow // to select the previous item
		
	on ctrl up:
		if sawCtrlTab:
			sawCtrlTab = false // reset the flag so we detect the next ctrl-tab down
			if the popup has been open < 450ms:
				sleep 200ms // so the JS has enough time to open the port from the menu
				send ctrl-Q // close the menu
			else if the popup is active:
				send enter // selects the tab and closes the popup

You'll probably also want to listen for ctrl-up/down and ctrl-shift-up/down and send just the up/down keys, so that you can also move the selection with the arrow keys while keeping ctrl down, and then release it to switch to the selected tab.

Hopefully that helps get you started.  It was mostly trial and error for the AHK code.  The tricky thing was tracking the state of the popup so you can send the right keys when ctrl-tab is pressed.

**Comment by jessecrouch** [comment]
Ah ok that makes way more sense. I wondered what all the delays and looping were about. Ok, I'll play with this and see what I can do. Thank you for breaking that down!

**Comment by fwextensions** [comment]
Yeah, they're mostly a workaround, as I don't know of a way in AHK to listen for a Chrome popup opening, so it has to poll and check the active window's width.  And there are timing issues if it doesn't wait for the menu to fully load before sending keystrokes.

**Comment by fwextensions** [comment]
@jessecrouch, curious if you were able to get Hammerspoon working for you.

**Comment by MartinLichtblau** [comment]
@jessecrouch What do you want to achieve with that script?

**Comment by jessecrouch** [comment]
@fwextensions @MartinLichtblau Way late replying to this sorry. I was trying to fix Most Recently Used tab in chrome (ctrl+tab to switch to most recently used rather than next tab - the way that alt+tab works).

Ultimately I have switched to Vivaldi browser which has native MRU and for keyboard I went to Karabiner Elements on mac and I'm using Autohotkey via EPKL (https://github.com/DreymaR/BigBagKbdTrixPKL) on WIndows.
